src/wallet/wallet.cpp

```diff
@@ -1457,6 +1512,50 @@ mapSproutNoteData_t CWallet::FindMySproutNotes(const CTransaction &tx) const
     return noteData;
 }
 
+
+/**
+ * Finds all output notes in the given transaction that have been sent to
+ * SaplingPaymentAddresses in this wallet.
+ *
+ * It should never be necessary to call this method with a CWalletTx, because
+ * the result of FindMySaplingNotes (for the addresses available at the time) will
+ * already have been cached in CWalletTx.mapSaplingNoteData.
+ */
+mapSaplingNoteData_t CWallet::FindMySaplingNotes(const CTransaction &tx) const
+{
+    LOCK(cs_SpendingKeyStore);
+    uint256 hash = tx.GetHash();
+
+    mapSaplingNoteData_t noteData;
+
+    // Protocol Spec: 4.19 Block Chain Scanning (Sapling)
+    int index = 0;
+    for (const OutputDescription &output : tx.vShieldedOutput) {
+        // FIXME: we need to iterate over all our spendingkeys
+        uint256 ivk;
+        auto result = SaplingNotePlaintext::decrypt(output.encCiphertext, ivk, output.ephemeralKey);
+        // FIXME: do we need to catch an exception when calling decrypt?
+        if (result) {
+            // TODO: we can compute the nullifier if we want to cache it...?
+            // auto pt = result.get();
+            // auto note = pt.note(ivk);
+            // boost::optional<uint256> nullifier(const SaplingSpendingKey &sk, const uint64_t position) const;
```

Comment:
See the first commit in #3417, where I alter `SaplingNote::nullifier()` to take a `SaplingFullViewingKey` (which is all that is needed). If we decide to not cache nullifiers, then we can punt on doing this here.

Reply:
In meeting we decided to not cache nullifiers given that we need the position.